### PR TITLE
[EraVM] Add EraVM C-API to LLD

### PR DESCRIPTION
**Pull request by PavelKopyl**
# Code Review Checklist

## Requirements
- [ ] Have the requirements been met?
- [ ] Have stakeholder(s) approved the change?

## Implementation
- [ ] Does this code change accomplish what it is supposed to do?
- [ ] Can this solution be simplified?
- [ ] Does this change add unwanted compile-time or run-time dependencies?
- [ ] Could an additional framework, API, library, or service improve the solution?
- [ ] Could we reuse part of LLVM instead of implementing the patch or a part of it?
- [ ] Is the code at the right abstraction level?
- [ ] Is the code modular enough?
- [ ] Can a better solution be found in terms of maintainability, readability, performance, or security?
- [ ] Does similar functionality already exist in the codebase? If yes, why isn’t it reused?
- [ ] Are there any best practices, design patterns or language-specific patterns that could substantially improve this code? 

## Logic Errors and Bugs
- [ ] Can you think of any use case in which the
code does not behave as intended?
- [ ] Can you think of any inputs or external events
that could break the code?

## Error Handling and Logging
- [ ] Is error handling done the correct way?
- [ ] Should any logging or debugging information
be added or removed?
- [ ] Are error messages user-friendly?
- [ ] Are there enough log events and are they
written in a way that allows for easy
debugging?

## Maintainability
- [ ] Is the code easy to read?
- [ ] Is the code not repeated (DRY Principle)?
- [ ] Is the code method/class not too long?

## Dependencies
- [ ] Were updates to documentation, configuration, or readme files made as required by this change?
- [ ] Are there any potential impacts on other parts of the system or backward compatibility?

## Security
- [ ] Does the code introduce any security vulnerabilities?

## Performance
- [ ] Do you think this code change decreases
system performance?
- [ ] Do you see any potential to improve the
performance of the code significantly?

## Testing and Testability
- [ ] Is the code testable?
- [ ] Have automated tests been added, or have related ones been updated to cover the change?
	- [ ] For changes to mutable state
- [ ] Do tests reasonably cover the code change (unit/integration/system tests)? 
	- [ ] Line Coverage
	- [ ] Region Coverage
	- [ ] Branch Coverage
- [ ] Are there some test cases, input or edge cases
that should be tested in addition?

## Readability
- [ ] Is the code easy to understand?
- [ ] Which parts were confusing to you and why?
- [ ] Can the readability of the code be improved by
smaller methods?
- [ ] Can the readability of the code be improved by
different function, method or variable names?
- [ ] Is the code located in the right
file/folder/package?
- [ ] Do you think certain methods should be
restructured to have a more intuitive control
flow?
- [ ] Is the data flow understandable?
- [ ] Are there redundant or outdated comments?
- [ ] Could some comments convey the message
better?
- [ ] Would more comments make the code more
understandable?
- [ ] Could some comments be removed by making the code itself more readable?
- [ ] Is there any commented-out code?
- [ ] Have you run a spelling and grammar checker?

## Documentation
- [ ] Is there sufficient documentation?
- [ ] Is the ReadMe.md file up to date?

## Best Practices
- [ ] Follow Single Responsibility principle?
- [ ] Are different errors handled correctly?
- [ ] Are errors and warnings logged?
- [ ] Magic values avoided?
- [ ] No unnecessary comments?
- [ ] Minimal nesting used?

## Experts' Opinion
- [ ] Do you think a specific expert, like a security
expert or a usability expert, should look over
the code before it can be accepted?
- [ ] Will this code change impact different teams, and should they review the change as well?

### CODE REVIEW

File: lld/Common/DriverDispatcher.cpp [hunk 1]
```diff
@@ -201,3 +246,34 @@ Result lld::lldMain(llvm::ArrayRef<const char *> args,
...
+                          llvm::raw_ostream &stdoutOS,
+                          llvm::raw_ostream &stderrOS,
+                          llvm::ArrayRef<DriverDefMemBuf> drivers) {
+  int r = 0;
+  {
+    // The crash recovery is here only to be able to recover from arbitrary
+    // control flow when fatal() is called (through setjmp/longjmp or
+    // __try/__except).
+    llvm::CrashRecoveryContext crc;
+    if (!crc.RunSafely([&]() {
```
Comment (akiramenai) [inline lld/Common/DriverDispatcher.cpp:263]: All pointers or references, so can be passed by value.

File: lld/Common/DriverDispatcher.cpp [hunk 2]
```diff
@@ -201,3 +246,34 @@ Result lld::lldMain(llvm::ArrayRef<const char *> args,
...
+          r = unsafeLldMainMemBuf(inBuffers, outBuffer, args, stdoutOS,
+                                  stderrOS, drivers, /*exitEarly=*/false);
+        }))
+      return {crc.RetCode, /*canRunAgain=*/false};
+  }
+
+  // Cleanup memory and reset everything back in pristine condition. This path
+  // is only taken when LLD is in test, or when it is used as a library.
+  llvm::CrashRecoveryContext crc;
+  if (!crc.RunSafely([&]() { CommonLinkerContext::destroy(); })) {
```
Comment (akiramenai) [inline lld/Common/DriverDispatcher.cpp]: no capture seems to be needed

File: lld/ELF/Driver.cpp [hunk 3]
```diff
@@ -1750,7 +1797,14 @@ void LinkerDriver::createFiles(opt::InputArgList &args) {
       break;
     }
     case OPT_script:
-      if (std::optional<std::string> path = searchScript(arg->getValue())) {
+      // EraVM local begin
+      if (config->useIOMemoryBuffers) {
+        if (std::optional<MemoryBufferRef> mb = readFile(arg->getValue()))
```
Comment (akiramenai) [inline lld/ELF/Driver.cpp:1802]: It looks like we want to avoid using `searchScript`. Could you explain the reason for that?

File: llvm/lib/Support/FileOutputBuffer.cpp [hunk 4]
```diff
@@ -109,6 +109,43 @@ class InMemoryBuffer : public FileOutputBuffer {
...
+
+  Error commit() override {
+    OutStream << StringRef(static_cast<const char *>(Buffer.base()),
+                           BufferSize);
+    OutStream.flush();
+    return Error::success();
+  }
+
+private:
+  // Buffer may actually contain a larger memory block than BufferSize
```
Comment (akiramenai) [inline llvm/lib/Support/FileOutputBuffer.cpp:143]: Do you mean that owning memory block size >= BifferSize?

**Review by akiramenai — COMMENTED** [review]
The patch is in a draft state but you mentioned it in the overview of the linker work, so I reviewed. Could you clarify what is left to be done?

**Comment by PavelKopyl** [comment]
> The patch is in a draft state but you mentioned it in the overview of the linker work, so I reviewed. Could you clarify what is left to be done?

I've switched it for ready to review. I changed LLD C-API to support metadata of arbitrary length (8b2002fc159e937cddb8767c4def9ea5948ac91e).

File: lld/Common/DriverDispatcher.cpp (hunk 1 above)
Comment (PavelKopyl) [reply lld/Common/DriverDispatcher.cpp:263]: You are right, but answering to this and to the following questions, the functions added to DriverDispatcher.cpp are almost copies of existing functions. It seems reasonable for me to preserver original structure. What do you think?

File: lld/Common/DriverDispatcher.cpp (hunk 2 above)
Comment (PavelKopyl) [reply lld/Common/DriverDispatcher.cpp]: Please, check my comment above.

File: lld/ELF/Driver.cpp (hunk 3 above)
Comment (PavelKopyl) [reply lld/ELF/Driver.cpp:1802]: You are right, we do not need `searchScript` when LLD In/Out is performed via memory buffers. Instead, we call   `readFile()` that contains a bit hackery code that instead of reading a file on OS just returns the reference to the passed memory buffer:
  ```
  ...
MemoryBufferRef mbref;
  if (config->useIOMemoryBuffers) {
    unsigned idx = 0;
    if (path.getAsInteger(10, idx)) {
      error(path + ": path should be an index");
      return std::nullopt;
    }
    if (idx >= config->inBuffers.size()) {
      error("memory buffer index is out of range");
      return std::nullopt;
    }
    mbref = config->inBuffers[idx];
    ...
```
where `arg->getValue()` is the memory buffer index, "0", or "1", so on.
Please, check the changes in `readFile()` in lld/ELF/InputFiles.cpp, line 234.

File: llvm/lib/Support/FileOutputBuffer.cpp (hunk 4 above)
Comment (PavelKopyl) [reply llvm/lib/Support/FileOutputBuffer.cpp:143]: Yes. Please note this class is almost a copy of the `InMemoryBuffer` one that has the same comment. The meaning of this comment arises from the implementation of an interface from `llvm/include/llvm/Support/Memory.h`.  I guess the allocated memory region can be more than `BifferSize` because of alignment requirement of the starting buffer address.

File: lld/Common/DriverDispatcher.cpp (hunk 1 above)
Comment (akiramenai) [reply lld/Common/DriverDispatcher.cpp:263]: I leave it up to you.
I'm just more nervous when reading a lambda that captures by reference as it should put a C++ programmer in a state of higher awareness. The lambda doesn't escape the scope and actually captures nothing so this awareness returns back to the normal level almost immediately. Still I expect us to read our local changes more frequently than the rest of the lld code, so stylistic consistency wouldn't be a priority for me.

Comment (PavelKopyl) [reply lld/Common/DriverDispatcher.cpp:263]: OK, agreed. Will fix it.

Comment (PavelKopyl) [reply lld/Common/DriverDispatcher.cpp:263]: Sorry, I've just have looked again. At least `stdoutOS`, `stderrOS` and `r` cannot be captured by value. 
Do you mean to employ default-capture by value and these three objects by reference, i.e `[=, &stdoutOS, &stderrOS, &r]`?

File: lld/Common/DriverDispatcher.cpp (hunk 2 above)
Comment (PavelKopyl) [reply lld/Common/DriverDispatcher.cpp]: You are right not needed. Removed.

**Comment by akiramenai** [comment]
Regression tests passed locally, so I merge.
